pkg/controller/imagelistpulljob/imagelistpulljob_controller.go

```diff
@@ -0,0 +1,425 @@
+// and what is in the ImageListPullJob.Spec
+// Automatically generate RBAC rules to allow the Controller to read and write ImageListPullJob
+func (r *ReconcileImageListPullJob) Reconcile(_ context.Context, request reconcile.Request) (res reconcile.Result, err error) {
+	klog.V(5).Infof("Starting to process ImageListPullJob %v", request.NamespacedName)
+
+	// 1.Fetch the ImageListPullJob instance
+	job := &appsv1alpha1.ImageListPullJob{}
+	err = r.Get(context.TODO(), request.NamespacedName, job)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			scaleExpectations.DeleteExpectations(request.String())
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// If resourceVersion expectations have not satisfied yet, just skip this reconcile
+	resourceVersionExpectations.Observe(job)
+	if isSatisfied, unsatisfiedDuration := resourceVersionExpectations.IsSatisfied(job); !isSatisfied {
+		if unsatisfiedDuration >= expectations.ExpectationTimeout {
+			klog.Warningf("Expectation unsatisfied overtime for %v, timeout=%v", request.String(), unsatisfiedDuration)
+			return reconcile.Result{}, nil
+		}
+		klog.V(4).Infof("Not satisfied resourceVersion for %v", request.String())
+		return reconcile.Result{RequeueAfter: expectations.ExpectationTimeout - unsatisfiedDuration}, nil
+	}
+
+	// delete
+	if job.DeletionTimestamp != nil {
+		resourceVersionExpectations.Delete(job)
+		return reconcile.Result{}, nil
+	}
+
+	// The Job has been finished
+	if job.Status.CompletionTime != nil {
+		var leftTime time.Duration
+		if job.Spec.CompletionPolicy.TTLSecondsAfterFinished != nil {
+			leftTime = time.Duration(*job.Spec.CompletionPolicy.TTLSecondsAfterFinished)*time.Second - time.Since(job.Status.CompletionTime.Time)
+			if leftTime <= 0 {
+				klog.Infof("Deleting ImageListPullJob %s/%s for ttlSecondsAfterFinished", job.Namespace, job.Name)
+				if err = r.Delete(context.TODO(), job); err != nil {
+					return reconcile.Result{}, fmt.Errorf("delete ImageListPullJob error: %v", err)
+				}
+				return reconcile.Result{}, nil
+			}
+		}
+		return reconcile.Result{RequeueAfter: leftTime}, nil
+	}
+
+	if scaleSatisfied, unsatisfiedDuration, scaleDirtyImagePullJobs := scaleExpectations.SatisfiedExpectations(request.String()); !scaleSatisfied {
+		if unsatisfiedDuration >= expectations.ExpectationTimeout {
+			klog.Warningf("Expectation unsatisfied overtime for ImageListPullJob %v, scaleDirtyImagePullJobs=%v, overtime=%v", request.String(), scaleDirtyImagePullJobs, unsatisfiedDuration)
+			return reconcile.Result{}, nil
+		}
+		klog.V(4).Infof("Not satisfied scale for ImageListPullJob %v, scaleDirtyImagePullJobs=%v", request.String(), scaleDirtyImagePullJobs)
+		return reconcile.Result{RequeueAfter: expectations.ExpectationTimeout - unsatisfiedDuration}, nil
+	}
+
+	job = job.DeepCopy()
```

Comment:
it is not necessary to deepcopy here, since objects from Get() is already deep copied